### PR TITLE
Initial user integration tests.

### PR DESCRIPTION
**Pull request by jrbenny35**
This adds some user integration tests to aide the SoftVision team a bit. Right now I have 3 tests.

```test_upload```: This will create a file and make sure it uploads by verifying a file uploads and is assigned a URL.

```test_download```: This will create a file, upload it and then download it making sure it is the same filename that was uploaded. We can expand this later to maybe check the sizes and such.

```test_progress```: This will create a file and make sure the progress bar shows up after it begins uploading.

These are python tests and use [Pipenv](https://docs.pipenv.org/) to manage dependencies as well as [tox](https://tox.readthedocs.io/en/latest/) as the virtualenv manager, and finally [pytest](https://docs.pytest.org/en/latest/) as the test runner. 

The tests build a docker container with the app and redis. I have a docker-compose override to expose the development port. 

I also added a URL for accessing the server besides ```localhost```, it is ```send.firefox.local```. This could be changed but it is needed to access the server within the selenium-firefox docker container.

I'll add some docs explaining more in the next PR, just wanted to get this up.

### CODE REVIEW

**Comment by dannycoates** [comment]
It's nice to see these ui tests running!

Unfortunately, I'm (frankly) not a python fan and the threat of maintaining a python environment and tests in this otherwise pristine js project steals my lunch. The code here is all very nice, readable, and easy to understand, but what happens when *I* try to change something? I'd be **much** more comfortable if they were in js. I'm not saying we can't add python even if it's the best option (I'll get over it), I'd just prefer to use js if we can. Adding another language also makes it one notch more difficult for contributors as well. Who would be responsible for maintaining this and adding new tests?

Regardless of language, I wonder if we could speed it up and maybe simplify the setup for easier maintenance with some changes.

The first is docker. Why are we building the docker image instead of running the server from the host? Is it to test the docker image? If so, I think that's not as important as running faster. Image problems should be very rare and can be easily caught on the dev or stage servers.

We have a saucelabs account (I started dabbling with it a few weeks ago but not on circleci). Should we run the tests against that instead of installing Firefox locally? IDK if it'd run faster, but it'd be less setup. We could also test against other browsers that way.

**Comment by jrbenny35** [comment]
> It's nice to see these ui tests running!
Unfortunately, I'm (frankly) not a python fan and the threat of maintaining a python environment and tests in this otherwise pristine js project steals my lunch. The code here is all very nice, readable, and easy to understand, but what happens when I try to change something? I'd be much more comfortable if they were in js. I'm not saying we can't add python even if it's the best option (I'll get over it), I'd just prefer to use js if we can. Adding another language also makes it one notch more difficult for contributors as well. Who would be responsible for maintaining this and adding new tests?

I would be adding new tests and maintenance. I am sure it could be done in js but the test-engineering team mainly uses python and we have a lot of tools that work well for it. If you change something and the test fails, I could add the fix, or show you how. The Object Model we use is very simple, and python is easy to read. If anything 'breaks' it would just be a locator or something like that, unless of course the project changed databases or something like that.

> Regardless of language, I wonder if we could speed it up and maybe simplify the setup for easier maintenance with some changes. The first is docker. Why are we building the docker image instead of running the server from the host? Is it to test the docker image? If so, I think that's not as important as running faster. Image problems should be very rare and can be easily caught on the dev or stage servers.

I used the docker image because it allowed me issue one command to start both the app and the backend and have them running as they would in production (or as close as possible). 

> We have a saucelabs account (I started dabbling with it a few weeks ago but not on circleci). Should we run the tests against that instead of installing Firefox locally? IDK if it'd run faster, but it'd be less setup. We could also test against other browsers that way.

Sauce labs is doable but for such a small project it isn't worth it. When saucelabs is down, or has a problem all the testing will fail/stop. 

We could also run the tests on Chrome, but once again, this is quite a small amount of tests currently. In the future it could be looked into I think.

**Comment by dannycoates** [comment]
> I would be adding new tests and maintenance

I appreciate the support but I also think it's important that anyone who can contribute to the project be able to write or change their own tests. A multi-language project significantly increases the barrier to entry and is harder to maintain.

I understand that you've got a lot of experience with python on your team and it makes your life easier when you don't need to switch languages per project, but that argument isn't any different from the one I'm making for javascript to you.

I have an alternative PR for selenium testing in javascript in #806

**Comment by jrbenny35** [comment]
@dannycoates I understand what you're saying but there have been no Integration tests added since this project began. If someone wanted to commit tests that bad I assume they would've by now. 

I am willing to write and maintain all the tests and dependencies. If you are willing to do that with JS than I suggest you merge the other PR and close this, but I am assuming that is something you don't want to do. I don't see why what language the tests are written in is a barrier if we get the coverage we need.

**Comment by jrbenny35** [comment]
I can tell you now I already have a contributor waiting to write tests for this project if they are in python. And there will be plenty more.

**Comment by dannycoates** [comment]
> I don't see why what language the tests are written in is a barrier if we get the coverage we need.

It is a barrier for new features an maintenance of broken tests. As soon as we add tests they will get broken and need to keep up with new features. Just because there haven't been any before doesn't mean we don't need to consider the maintenance burden going forward. It's *even more important* that any contributor can maintain them otherwise the tests stagnate.

**Comment by dannycoates** [comment]
@clouserw gave me some additional context about this that I was unaware of before. It makes sense to use python if it makes integrating with the larger mozilla testing ecosystem possible.

It remains important to me that folks contributing to the codebase be able to easily run the integration tests locally. It'd be great if they could run via a docker container as easily as the POC in #806. At the very least we need documentation for how to setup and run them.

**Comment by dannycoates** [comment]
Here's a few things I'd like to see:

- All files related to integration tests be under `test/integration` including `tox.ini` `Pipfile` etc.
- Circleci shouldn't build the app docker image, it should just run `npm start` or equivalent
- It'd be nice if all the python setup and execution could be contained in a dockerfile, ala `test/integration/Dockerfile`
- The end goal is that there's nothing special about running the integration tests on circleci vs local. It should all roll up into an npm script runnable like, `npm run test:integration`

I can help with wiring up the execution side if you can make a dockerfile that installs all the python dependencies

**Comment by dannycoates** [comment]
This is incomplete, but a start at running the tests inside docker, that you might be able to build on.

https://github.com/mozilla/send/commit/f32d685cb23e079ed393ba86c4c4baf379c2d0f3

From the `int` branch if you run `docker-compose up` in the `test/integration` directory you should see it build a new container and run the tests against the dev server. With some more config it should be able to run against localhost.

**Comment by jrbenny35** [comment]
> This is incomplete, but a start at running the tests inside docker, that you might be able to build on.
From the int branch if you run docker-compose up in the test/integration directory you should see it build a new container and run the tests against the dev server. With some more config it should be able to run against localhost.

This is really good and super helpful. I will definitely use this

**Comment by dannycoates** [comment]
> I need all assets and files that would be available to have the server running as close to production. Just using the docker file and building it that way, the .dockerignore doesn't allow for any of those other files I need to be added to the image.

The production image is built using that .dockerignore file. Nothing excluded by it is in production, so I don't understand why you think you need more. All the assets are the same whether the server is started with `npm start` or `npm run build && npm run prod`, just that the former watches for source changes and recompiles. Doing integration tests against either will yield the exact same results.

**Comment by jrbenny35** [comment]
I am not talking about using the npm scripts. Just using a ```docker-compose up -d``` command. I get this: 
```
Error: Cannot find module '../dist/undefined'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.serverTranslator [as getTranslator] (/app/common/locales.js:21:10)
    at module.exports (/app/server/state.js:9:24)
    at index (/app/server/routes/pages.js:13:47)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at next (/app/node_modules/express/lib/router/route.js:137:13)
    at module.exports (/app/server/middleware/language.js:43:3)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at next (/app/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at /app/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at jsonParser (/app/node_modules/body-parser/lib/types/json.js:109:7)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at /app/server/routes/index.js:52:5
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at hsts (/app/node_modules/hsts/index.js:42:5)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at internalNext (/app/node_modules/helmet/index.js:45:33)
    at xXssProtection (/app/node_modules/x-xss-protection/index.js:26:7)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at nosniff (/app/node_modules/dont-sniff-mimetype/index.js:4:5)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at ienoopen (/app/node_modules/ienoopen/index.js:4:5)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at hsts (/app/node_modules/hsts/index.js:42:5)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at hidePoweredBy (/app/node_modules/hide-powered-by/index.js:12:7)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at frameguard (/app/node_modules/frameguard/index.js:38:5)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at dnsPrefetchControl (/app/node_modules/dns-prefetch-control/index.js:10:7)
    at internalNext (/app/node_modules/helmet/index.js:49:7)
    at helmet (/app/node_modules/helmet/index.js:52:5)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at expressInit (/app/node_modules/express/lib/middleware/init.js:40:5)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at query (/app/node_modules/express/lib/middleware/query.js:45:5)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/app/node_modules/express/lib/router/index.js:317:13)
    at /app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/app/node_modules/express/lib/router/index.js:335:12)
    at next (/app/node_modules/express/lib/router/index.js:275:10)
    at Function.handle (/app/node_modules/express/lib/router/index.js:174:3)
    at Function.handle (/app/node_modules/express/lib/application.js:174:10)
    at Server.app (/app/node_modules/express/lib/express.js:39:9)
    at emitTwo (events.js:126:13)
    at Server.emit (events.js:214:7)
    at parserOnIncoming (_http_server.js:619:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:23)
```

It seems like it is because I need some files within the ```.dockerignore```. I don't know how this is deployed but isn't there files stored on the server or through a CDN or something like that? Please excuse my ignorance but I have not much experience with deployments and such.

File: package.json [hunk 1]
```diff
@@ -28,6 +28,9 @@
     "test": "npm-run-all test:*",
     "test:backend": "nyc mocha --reporter=min test/backend",
     "test:frontend": "cross-env NODE_ENV=development node test/frontend/runner.js && nyc report --reporter=html",
+    "test:integration-local": "tox -c test/integration/tox.ini",
+    "test:integration-docker": "docker-compose exec -T --user root selenium-firefox tox -c test/integration/tox.ini",
```
Comment (dannycoates) [inline package.json]: Both `test:integration-local` and `test:integration-docker` shouldn't run when `npm test` is run. For now I'd like  neither of them to run on `npm test` until we can go from `npm install` to `npm test` without any intermediate steps. I'm happy to do that later. All this entails is renaming them to not start with `test:`

File: test/integration/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+[git-clone]: https://help.github.com/articles/cloning-a-repository/
+[git-fork]: https://help.github.com/articles/fork-a-repo/
+[geckodriver]: https://github.com/mozilla/geckodriver/releases/tag/v0.19.1
+[pypom]: http://pypom.readthedocs.io/en/latest/
+[pytest]: https://docs.pytest.org/en/latest/
+[pytest-html]: https://github.com/pytest-dev/pytest-html
+[pytest-selenium]: http://pytest-selenium.readthedocs.org/
+[Selenium]: http://selenium-python.readthedocs.io/index.html
+[selenium-api]: http://selenium-python.readthedocs.io/locating-elements.html
+[Tox]: http://tox.readthedocs.io/
```
Comment (dannycoates) [inline test/integration/README.md]: Thank you, this doc is very helpful.

File: package.json (hunk 1 above)
Comment (jrbenny35) [reply package.json]: Something like ```test-``` ?

File: circle.yml [hunk 3]
```diff
@@ -58,6 +98,7 @@ workflows:
           filters:
             branches:
               ignore: master
+      - integration_tests
```
Comment (jrbenny35) [inline circle.yml]: I think I have to move this as you don't want these to run on every PR correct?

Comment (dannycoates) [reply circle.yml]: Let's leave it and see how it goes

File: package.json (hunk 1 above)
Comment (dannycoates) [reply package.json]: sure

**Comment by jrbenny35** [comment]
Any other changes I should add?

**Comment by dannycoates** [comment]
Looks good. Thanks @jrbenny35 😄

**Pull request merged**